### DISCUSSION TITLE
AgentGroupChat with AzureAssistantAgent agents and state management

### DISCUSSION THREAD

**Discussion by vslepakov** [microsoft/semantic-kernel#10041] [Q&A] [answered]
## Question: Reusing `AzureAssistantAgent` in `AgentGroupChat`

I have the following setup:

- An API that accepts user input
- It uses `AgentGroupChat` to orchestrate several `AzureAssistantAgent` instances
- Currently, I'm creating a new instance of `AgentGroupChat` on each request
- I attempt to reuse the `AzureAssistantAgent` instances by calling `AzureAssistantAgent.retrieve` with an `agent_id` (the agents are initially created on the first request)

**Issue:**  
When I try to retrieve an existing agent by `agent_id`, I receive the error message:  "No assistant found with id ..."

**Questions:**  
- Is this error expected or by design?  
- What is the correct approach for managing state and concurrency (multithreading) in `AgentGroupChat` when working with `AzureAssistantAgent` instances?

**Comment by moonbox3** [comment, marked as answer]
Hi @vslepakov, thanks for your question. Are you handling the retrieval in the following manner?

First we create the assistant agent, as an example:

```python
agent = await AzureAssistantAgent.create(
    kernel=kernel,
    service_id=service_id,
    name=AGENT_NAME,
    instructions=AGENT_INSTRUCTIONS,
    enable_code_interpreter=True,
)
```

Then we can retrieve it using the assistant ID:

```python
assistant_id = agent.assistant.id

# Retrieve the agent using the assistant_id
retrieved_agent: AzureAssistantAgent = await AzureAssistantAgent.retrieve(
    id=assistant_id,
    kernel=kernel,
)

# Define a thread and invoke the agent with the user input
thread_id = await retrieved_agent.create_thread()

try:
    await invoke_agent(retrieved_agent, thread_id, "Tell me a joke about bears.")
finally:
    await agent.delete()
    await retrieved_agent.delete_thread(thread_id)
```

Code is from this sample (I realize we have a bug in the sample where it only retrieves the OpenAIAssistantAgent, I will fix this):

https://github.com/microsoft/semantic-kernel/blob/main/python/samples/concepts/agents/assistant_agent_retrieval.py

**Reply by vslepakov** [reply]
Hi @moonbox3, thanks for your answer! Looking at your code, I just realized that I am retrieving the `assistant_id` like so `agent.id` instead of  `agent.assistant.id`. Agent retrieval works now. 

It would be great to have samples for agent usage with `AgentGroupChat` in the context of a web api (multiple simultaneous requests, agent and agent group chat lifecycle, history handling in `AgentGroupChat` when retrieving existing assistant agents etc.)

Thanks!
